modules/markup/markdown/markdown_test.go

```diff
@@ -100,7 +100,7 @@ func testAnswers(baseURLContent, baseURLImages string) []string {
 <p>Ideas and codes</p>
 
 <ul>
-<li>Bezier widget (by <a href="` + AppURL + `r-lyeh" rel="nofollow">@r-lyeh</a>) <a href="http://localhost:3000/ocornut/imgui/issues/786" rel="nofollow">#786</a></li>
+<li>Bezier widget (by <a href="` + AppURL + `r-lyeh" rel="nofollow">@r-lyeh</a>) <a href="http://localhost:3000/ocornut/imgui/issues/786" rel="nofollow">ocornut/imgui#786</a></li>
```

Comment:
So we will miss a test?

Reply:
This just changes the test to match what would be the new behavior. markdown_test.go defines this at the top:

```
const AppURL = "http://localhost:3000/"
const Repo = "gogits/gogs"
```
So that link should now be rendered as org/repo#id since it is from a different repo than the current context. It still tests the same function, it just matches what would be the new output of that function in this case.

I did add an additional test for matching an issue when it is in the same repository to make sure that case is still included in the tests (https://github.com/go-gitea/gitea/pull/6273/commits/e85bd10e7332c9596b4e3e6a261de374ef98556c#diff-94ed3852444a1c36b7f65b28312bf01aR224)

I can add another one to this file too if you'd like. Thanks!

Reply:
While looking into adding some additional tests, I see some of the tests are run from different locations. 

Is there a better way (inside of modules/markup/html.go) to get the current organization name and repo name of a page somebody is on? Right now I'm trying to parse ctx.urlPrefix but thats a but unpredictable if the URL can vary based on what type of page the user is on. Thanks!

Reply:
Have you tried: `ctx.metas["user"] + "/" + ctx.metas["repo"]`

Reply:
I have tried that, but it appears those are only available if using an external issue tracker:

https://github.com/go-gitea/gitea/blob/f7ffb190c66319e9f2f0aaacaccd0bf40614eb12/models/repo.go#L473-L477

Though I'm not familiar enough with it all to be sure of that. It doesn't appear that the current user or repo name is available in this context but perhaps it should be for this purpose.

Reply:
Ok, I'd suggest changing that code to always provide the owner and the repo name.

Reply:
I tried this but it went down a small rabbit hole where there I believe there would actually need to be a change to the Markdown API at routers/api/v1/misc/markdown.go which doesn't seem ideal. Changing the above function in repo.go allows access to ctx.metas["repo"] etc when viewing/writing an issue, but not when the text is actually processed (which is what you'd want in this case).

However, I noticed other functions in html.go use setting.AppURL for reliability when parsing URLS and that does indeed seem to be a simpler solution that works more reliably and passes all current test. I've also added a 2nd test to modules/markup/markdown/markdown_test.go per @lunny comment above so it now tests both behaviors in the wiki context and nothing is lost there. Thanks much for the feedback sofar.

Reply:
Hmm...

I have a massive aversion to people doing url-munging by hand to be honest

1. Mostly people tend to get it wrong.  For example: 

https://github.com/go-gitea/gitea/blob/fabf3f2fc29e143dabefd504cda78d3f47807d2c/cmd/web.go#L60

Is wrong and has been for over a year - and that's with using go's URL package where it tells you what to do to do things correctly.

2. It strictly enforces an url-scheme that may or may not change in the future, and when these url-schemes change it's often impossible to find all the places that it has been relied on. 

Now, I don't think your url-munging is necessarily incorrect, and I think there's plenty of this kind of thing going on in Gitea in any case - I'm probably also responsible for some myself, but I think we should try to avoid doing this. Is there any way you think of to not have to do this?

Reply:
My Idea would be the same as yours which is to just have the repo and org name available in the current context -- I'm just not 100% how to make that happen after trying. I *believe* it is missing when rendering here:

https://github.com/go-gitea/gitea/blob/e836b8869bfb89fe7d3d1f898c35f77ffb6dfaa6/routers/api/v1/misc/markdown.go#L46-L55

That render function itself has some url-munging so be warned ;)

I just followed it there through a stack trace and am admittedly not familiar enough with the code or structure to say anything for certain -- that is just where the trail ended for my limited knowledge. If you know of a change that would allow access to the repo and org name when processing this text I'd be more than happy to discuss or test. Thanks!

Reply:
I took one more look at this and I think getting the repo/org name another way wouldn't be possible because of how this uses the API to render the content. In order to render issues it has to POST to the markdown api:
```
POST /api/v1/markdown
```
According to https://try.gitea.io/api/swagger#/miscellaneous/renderMarkdown it takes the following:
```
{
  "Context": "string",
  "Mode": "string",
  "Text": "string",
  "Wiki": true
}
```
So right now, this is passed as {{.Repo.RepoLink}} in all places, for example:
https://github.com/go-gitea/gitea/blob/b3ad5eb414c67f96594a2863b4ada909cf2eea7f/templates/repo/issue/comment_tab.tmpl#L7

That is then used to build what ends up the urlPrefix:
https://github.com/go-gitea/gitea/blob/e836b8869bfb89fe7d3d1f898c35f77ffb6dfaa6/routers/api/v1/misc/markdown.go#L48

So I think it is the best we can get here to just split that URL apart again reversing how it was put together from Repo.RepoLink in the first place (what this PR currently does).

Reply:
Great! So results of those calls aren't going to work when we have an external issue tracker! That means all our previews are incorrect.  They're actually incorrect on any sub-mounted gitea anyway! (RepoLink should have the AppSubUrl on it already and then it has AppUrl prepended to it so if your gitea is proxied to `https://www.example.com/git` then you end up with `https://www.example.com/git/git/user/repo`!)

I guess the point of this api link was that you could use it to render random markdown without giving it a complete context - but we can't do that - context is key to rendering documents, especially when you have context dependent links.

Context should be determined as close as possible to point it can be - that means that `api/v1/misc/markdown.go` should decompose the context url and determine if it matches a repository url and then create a valid meta context for that - if we want to avoid calls to the database we should pass in a meta context properly. (Although that means a change to the sdk - it can be non-breaking by just trying to work out the context.)

1. The url munging should happen in `/api/v1/misc/markdown.go`
2. If it's a repository url -> Get the repository and get its meta context
3. Change the repository metaContext to always return something useful
4. Use the meta context in the renderer.
5. If there is no meta context - then don't render the issue links - we're not in a repo.

Reply:
Take a look at my changes in https://github.com/zeripath/gitea/tree/improve-issue-links

Reply:
Awesome, that looks great to me! I can see the metas now available in the proper context. If you want I can merge this into my branch and update this PR to use ctx.metas["X"]. Or if you want to make a separate PR for these changes I can update this one after they are approved.

Also for the urlPrefix part, perhaps it could use the existing url matching library
```
linkRegex, _ := xurls.StrictMatchingScheme("https?://")
m := linkRegex.FindStringIndex(urlPrefix)
if m == nil {
    urlPrefix = util.URLJoin(setting.AppURL, form.Context)
}
```
Which while perhaps a bit overkill does seem easy to understand and much more reliable compared to various net/url options as it already has all the logic of matching actual URLs and not things that look like URLs, paths, etc.. I tried using the more std url.ParseRequestURI but it accepts paths as valid as well so considers the current situation of zeripath/gitea a valid URI :-/

Reply:
Yeah pull 'em. I wouldn't have put them up if I didn't want you to pull them!

Using xurls sounds like a good idea too.